### PR TITLE
comp: selector: add support for binary configuration data

### PR DESCRIPTION
**Pull request by lbetlej**
Channel selector component extended with support for configuration get/set using
COMP_CMD_GET_DATA & COMP_CMD_SET_DATA commands of SOF_CTRL_CMD_BINARY type.

### CODE REVIEW

File: src/audio/selector.c [hunk 1]
```diff
@@ -194,13 +194,30 @@ static int selector_params(struct comp_dev *dev)
...
+		trace_selector("selector_ctrl_set_data(), SOF_CTRL_CMD_BINARY");
+
+		cfg = (struct sof_sel_config *)cdata->data->data;
+		/* Just copy the configuration & verify input params.*/
+		ret = sel_set_channel_values(cd, cfg->in_channels_count,
+				  cfg->out_channels_count, cfg->sel_channel);
+		break;
+	default:
+		trace_selector_error("selector_ctrl_set_cmd() error: "
+		"invalid cdata->cmd = %u", cdata->cmd);
```
Comment (tlauda) [inline src/audio/selector.c]: Wrong indentation.

File: src/audio/selector.c [hunk 2]
```diff
@@ -233,17 +268,27 @@ static int selector_cmd(struct comp_dev *dev, int cmd, void *data,
 			int max_data_size)
 {
 	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
 
 	trace_selector("selector_cmd()");
 
 	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = selector_ctrl_set_data(dev, cdata);
```
Comment (tlauda) [inline src/audio/selector.c]: Break missing.

File: src/audio/selector.c [hunk 3]
```diff
@@ -233,17 +268,27 @@ static int selector_cmd(struct comp_dev *dev, int cmd, void *data,
...
+	case COMP_CMD_GET_DATA:
+		ret = selector_ctrl_get_data(dev, cdata, max_data_size);
```
Comment (tlauda) [inline src/audio/selector.c]: Break missing.

File: test/cmocka/src/audio/selector/selector_test.c [hunk 4]
```diff
@@ -78,17 +78,17 @@ static int setup(void **state)
...
+	cd->config.sel_channel = parameters->sel_channel;
 
 	cd->sel_func = sel_get_processing_function(sel_state->dev);
 
 	/* allocate new sink buffer */
 	sel_state->sink = test_malloc(sizeof(*sel_state->sink));
 	sel_state->dev->params.frame_fmt = parameters->sink_format;
 	size = parameters->frames * comp_frame_bytes(sel_state->dev);
-	if(cd->out_channels_count == SEL_SINK_1CH){
+	if(cd->config.out_channels_count == SEL_SINK_1CH){
```
Comment (tlauda) [inline test/cmocka/src/audio/selector/selector_test.c]: Missing spaces.

File: src/audio/selector.c [hunk 5]
```diff
@@ -194,13 +194,30 @@ static int selector_params(struct comp_dev *dev)
  * \param[in,out] cdata Control command data.
  * \return Error code.
  */
-static int selector_ctrl_set_cmd(struct comp_dev *dev,
+static int selector_ctrl_set_data(struct comp_dev *dev,
 				 struct sof_ipc_ctrl_data *cdata)
```
Comment (mrajwa) [inline src/audio/selector.c]: wrong indentation.

Comment (mrajwa) [inline src/audio/selector.c]: wrong indentation.

Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.c (hunk 3 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.c (hunk 2 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.c (hunk 1 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: test/cmocka/src/audio/selector/selector_test.c (hunk 4 above)
Comment (lbetlej) [reply test/cmocka/src/audio/selector/selector_test.c]: fixed

File: src/audio/selector.c [hunk 6]
```diff
@@ -210,15 +227,33 @@ static int selector_ctrl_set_cmd(struct comp_dev *dev,
...
 
-	return 0;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		trace_selector("selector_ctrl_get_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Copy back to user space */
+		if (memcpy_s(cdata->data->data, ((struct sof_abi_hdr *)
```
Comment (lgirdwood) [inline src/audio/selector.c]: This doesn't follow coding convention and it's possible checkpatch doesn't detect it either. it should be 
```
ret = memcpy();
if (ret < 0)
    return ret;
```

File: src/audio/selector.h [hunk 7]
```diff
@@ -63,22 +63,25 @@
...
+struct sof_sel_config {
 	/* selector supports 1 input and 1 output */
 	uint32_t in_channels_count;	/**< accepted values 2 or 4 */
 	uint32_t out_channels_count;	/**< accepted values 1 or 2 or 4 */
-
 	/* note: if 2 or 4 output channels selected the component works in
 	 * a passthrough mode
 	 */
 	uint32_t sel_channel;	/**< 0..3 */
+} __attribute__((packed));
```
Comment (lgirdwood) [inline src/audio/selector.h]: This wont need to be packed as it's not used outside of the FW.

File: src/audio/selector.c [hunk 8]
```diff
@@ -93,9 +93,9 @@ static int sel_set_channel_values(struct comp_data *cd, uint32_t in_channels,
 		return -EINVAL;
 	}
 
-	cd->in_channels_count = in_channels;
-	cd->out_channels_count = out_channels;
-	cd->sel_channel = ch_idx;
+	cd->config.in_channels_count = in_channels;
```
Comment (lgirdwood) [inline src/audio/selector.c]: This should really follow existing naming convention where we use source and sink instead of in and out respectively. This could be a separate patch though..

Comment (lbetlej) [reply src/audio/selector.c]: OK, I'll make it as a separate patch

File: src/audio/selector.c (hunk 6 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.h (hunk 7 above)
Comment (lbetlej) [reply src/audio/selector.h]: fixed
